### DISCUSSION TITLE
Improve build cache & installation for projects that use npm to install deps

### DISCUSSION THREAD

**Discussion by ghost** [vercel/vercel#6324] [Ideas]
Currently, the default install command for a next.js project on vercel is to use `npm install` or `yarn install`. Besides that, vercel also seems to cache `node_modules` for the build environment. I think this could be improved:

1. When caching for faster npm installs, it is generally recommended to cache `~/.npm` (on Posix) instead of `node_modules` as caching `node_modules` can break across Node versions and won't work with `npm ci` (see [here](https://github.com/actions/cache/blob/main/examples.md#node---npm)).
2. When using npm to install in an automated environment, [npm recommends using `npm ci`](https://docs.npmjs.com/cli/v7/commands/npm-ci#synopsis). I would expect vercel to stick to the recommendations of the upstream package manager. Besides that, `npm ci` won't update the lockfile if it's out of sync (which you don't want on CI/CD) and has the potential to be significantly faster.

You could say that running `npm i` in an automated environment is maybe not completely _horrible_. It generally works as expected. However, I don't think there are any reasons to _prefer_ `npm i` over `npm ci` in an automated environment. `npm ci` was specifically built for use in environments like the vercel deployment environment, so I'd vastly prefer vercel catering to the needs of `npm ci` in the vercel build environment. Currently however, because vercel caches `node_modules` and not `.npm`, the efficiency of `npm ci` is hindered, and I think it would be good to resolve that.

**Comment by paulogdm** [comment]
Thanks a ton for your message! I have relayed it internally.
cc @styfle

**Comment by styfle** [comment]
> caching node_modules can break across Node versions and won't work with npm ci

Vercel already handles this case and invalidates the build cache when the Node.js version changes.

> Currently however, because vercel caches node_modules and not .npm, the efficiency of npm ci is hindered, and I think it would be good to resolve that.

We would have to measure to compare the two because I'm not so sure this is the case. We can't remove `node_modules` caching because yarn and legacy npm still need it (remember the user can customize the Install Command).

**Reply by ghost** [reply]
Hi @paulogdm, @styfle, thanks for the quick reply! 

> Vercel already handles this case and invalidates the build cache when the Node.js version changes.

Nice, that's good to hear.

> We would have to measure to compare the two because I'm not so sure this is the case.

So what I was basing this statement on is the [`npm ci` docs](https://docs.npmjs.com/cli/v7/commands/npm-ci#description), where they mention: "_npm ci will be significantly faster when the node_modules folder is missing or empty._". Which makes sense, because `npm ci` will remove node_modules if present before installing, which will always slow it down compared to not having to remove anything.

> We can't remove node_modules caching because yarn and legacy npm still need it (remember the user can customize the Install Command).

I didn't touch on yarn, but for yarn `~/.cache/yarn` could be used (see the caching strategy this action uses: https://github.com/bahmutov/npm-install#basic).

For legacy npm: on my account it seems that vercel supports node 12 and 14. Node 12 comes with npm 6 I believe, which supports the .npm cache: https://docs.npmjs.com/cli/v6/commands/npm-cache (I'm not sure if there are any other npm versions available for use with the vercel build environment, outside of manual user installs of npm).

Now before we focus too much on purely the speed, I feel I should stress the following: npm install and yarn install aren't meant for usage in an automated environment. `npm ci` and `yarn --frozen-lockfile` are more appropriate, regardless of any speed benefits. From the yarn docs:

<img width="732" alt="Screenshot 2021-06-05 at 09 03 32" src="https://user-images.githubusercontent.com/7355199/120883387-196dc280-c5dd-11eb-8733-8970cdcfb083.png">

This goes for `npm ci` as well. It also won't update the lockfile if the lockfile is out of sync, which is what you want on ci. So `npm ci` and `yarn --frozen-lockfile` are the recommended commands for installing dependencies on ci.

**Reply by styfle** [reply]
Thanks, we'll look into caching global directories like `~/.npm`

**Reply by ghost** [reply]
@styfle, thanks I appreciate it! And thanks for all the good work done on the vercel platform!
